### DISCUSSION TITLE
Help: How exactly is the path to a node build?

### DISCUSSION THREAD

**Discussion by Silas598** [eclipse-milo/milo#1349] [General]
I am just getting started with OPC for an upcoming project and ran into some issues interacting with OPC.
I am most likely just missing some basic OPC knowledge, but I wasn't sure where to look for it, so any help or guidance is appreciated. 

We have a Demo PLC in our office that I can browse using  UaExpert:
![UaExpert-Structurepng](https://github.com/user-attachments/assets/60e0829b-6edc-4641-85be-4e28ad75870b)

And I wrote some code that allows me to browse children starting from "DataBlocksGlobal":

        namespaceIndex = 3;
        identifier = "DataBlocksGlobal";

        UaObjectNode myNode = opcUaClient.getAddressSpace().getObjectNode(
                new NodeId(namespaceIndex, identifier)
        );

        System.out.println(myNode.getNodeId());
        System.out.println(myNode.getBrowseName());

        List<? extends UaNode> childNodes = myNode.browseNodes();
        for (UaNode childNode : childNodes){
            System.out.println("--Child: " + childNode.getBrowseName());

            List<? extends UaNode> babyNodes = childNode.browseNodes();
            for(UaNode babyNode : babyNodes){
                QualifiedName name = babyNode.getBrowseName();
                System.out.println("----Baby: " + name);
                System.out.println("-------NodeId: " + babyNode.readAttribute(AttributeId.NodeId).getValue());
                System.out.println("-------Value: " + babyNode.readAttribute(AttributeId.Value).getValue());
            }

        }

This generates the following oputput:

        NodeId{ns=3, id=DataBlocksGlobal}
        QualifiedName{name=DataBlocksGlobal, namespaceIndex=3}
        --Child: QualifiedName{name=DB100_BOOL, namespaceIndex=3}
        ----Baby: QualifiedName{name=xBool_0, namespaceIndex=3}
        -------NodeId: Variant{value=NodeId{ns=3, id="DB100_BOOL"."xBool_0"}}
        -------Value: Variant{value=false}

This allows me to read the actual values in xBool_0 for example.

When I try to browse the node directly however, I get the error Bad_NodeIdUnknown:

         namespaceIndex = 3;
        //identifier = "DataBlocksGlobal                        //Works
        //identifier = "DataBlocksGlobal.DB100_BOOL.xBool_0";   //yields status=Bad_NodeIdUnknown, description=The node id refers to a node that does not exist in the server address space.
        //identifier = "xBool_0";                               //yields status=Bad_NodeIdUnknown, description=The node id refers to a node that does not exist in the server address space.

        AddressSpace addressSpace = opcUaClient.getAddressSpace();
        UaNode node = addressSpace.getObjectNode(
                new NodeId(namespaceIndex, identifier)
        );

So my question is: How exactly is the path to a node build?

**Comment by kevinherron** [comment]
Be more careful when you try to reconstruct these NodeIds. Siemens puts actual quotes into the String identifier.

e.g. 
```java
NodeId nodeId = NodeId.parse("ns=3;s=\"DB100_BOOL\".\"xBool_0\");
```
or
```java
NodeId nodeId = new NodeId(3, "\"DB100_BOOL\".\"xBool_0\"");
```

In general, though, you cannot rely on a NodeId having any relation to its position in the browse hierarchy. Some servers do use relatively predictable String-based NodeIds, but it's absolutely not a requirement.

**Reply by Silas598** [reply]
Thank you very much! I actually tried it with quotes but missed escaping them.

What would be a better way to access specific nodes, if not by their NodeId? Or do you just mean to browse like I did before until I find the NodeId as I cannot rely on the path being the same every time?

**Reply by kevinherron** [reply]
Accessing Nodes by their NodeId is the right way to do it. 

My point was that from server to server you probably can't intuit or guess the NodeIds, you have to browse to discover them, and even within a server there is not always a relationship between the position in the browse hierarchy you find a Node and the contents of its NodeId.
